### PR TITLE
Implement "conversation lock" for issue comments

### LINKED ISSUE #5063: Add "lock conversation" for issues
It would be cool if Gitea would support the "lock conversation" feature like github does.

### PR DESCRIPTION
**Pull request by adelowo**
The goal of this PR is to add a ___lock conversation___ feature 
present in Github.

Fixes https://github.com/go-gitea/gitea/issues/5063 .

- [x] Lock Conversation
- [x] Unlock Conversation
- [x] Update issues templates to show lock events
- [x] Disable entire comment box when an issue is locked if the user is not an ___instance admin___ or repo admin/writer
- [x] Allow a reason to be specified when locking an issue 

screenshot 2018-10-13 at 00 36 06

> Youtube video of what it looks like 

[Conversation lock](https://youtu.be/7sV9v3ld_8M)

### CODE REVIEW

**Comment by adelowo** [comment]
Any idea why drone is failing with `package github.com/golang/lint/golint: code in directory /go/src/github.com/golang/lint/golint expects import "golang.org/x/lint/golint"` ? `make lint` doesn't fail locally.. No idea why

Started failing at https://github.com/go-gitea/gitea/pull/5073/commits/9d66bc3581b566189f68f09513bc22274bfa7a40 

---- 
Edit

Looks like all builds are failing with this cryptic error, #5009 and https://github.com/go-gitea/gitea/pull/5051 just failed

File: custom/conf/app.ini.sample [hunk 1]
```diff
@@ -323,8 +327,8 @@ ENABLE_CAPTCHA = false
 CAPTCHA_TYPE = image
 ; Enable recaptcha to use Google's recaptcha service
 ; Go to https://www.google.com/recaptcha/admin to sign up for a key
-RECAPTCHA_SECRET  = 
-RECAPTCHA_SITEKEY = 
+RECAPTCHA_SECRET  =
```
Comment (jonasfranz) [inline custom/conf/app.ini.sample]: unrelated change

File: templates/repo/issue/comment_tab.tmpl [hunk 2]
```diff
@@ -11,7 +11,7 @@
 		{{.i18n.Tr "loading"}}
 	</div>
 </div>
-{{if .IsAttachmentEnabled}}
+{{if .IsAttachmentEnabled  }}
```
Comment (jonasfranz) [inline templates/repo/issue/comment_tab.tmpl]: unrelated change

File: templates/repo/issue/view_content/comments.tmpl [hunk 3]
```diff
@@ -1,7 +1,11 @@
 {{range .Issue.Comments}}
 	{{ $createdStr:= TimeSinceUnix .CreatedUnix $.Lang }}
 
-	<!-- 0 = COMMENT, 1 = REOPEN, 2 = CLOSE, 3 = ISSUE_REF, 4 = COMMIT_REF, 5 = COMMENT_REF, 6 = PULL_REF, 7 = COMMENT_LABEL, 12 = START_TRACKING, 13 = STOP_TRACKING, 14 = ADD_TIME_MANUAL, 16 = ADDED_DEADLINE, 17 = MODIFIED_DEADLINE, 18 = REMOVED_DEADLINE, 19 = ADD_DEPENDENCY, 20 = REMOVE_DEPENDENCY, 21 = CODE, 22 = REVIEW -->
+	<!-- 0 = COMMENT, 1 = REOPEN, 2 = CLOSE, 3 = ISSUE_REF, 4 = COMMIT_REF,
+	 5 = COMMENT_REF, 6 = PULL_REF, 7 = COMMENT_LABEL, 12 = START_TRACKING,
+	 13 = STOP_TRACKING, 14 = ADD_TIME_MANUAL, 16 = ADDED_DEADLINE, 17 =
+	MODIFIED_DEADLINE, 18 = REMOVED_DEADLINE, 19 = ADD_DEPENDENCY, 20 =
+	REMOVE_DEPENDENCY, 21 = CODE, 22 = REVIEW 23 = ISSUE_LOCKED 24 = ISSUE_LOCKED -->
```
Comment (jonasfranz) [inline templates/repo/issue/view_content/comments.tmpl]: Addd missing commas

File: templates/repo/issue/view_content/sidebar.tmpl [hunk 4]
```diff
@@ -335,6 +335,91 @@
 					</div>
 				{{end}}
 			</div>
+
+			{{ if .IsRepoAdmin }}
+			<div class="ui divider"></div>
+			<div class="ui watching">
```
Comment (jonasfranz) [inline templates/repo/issue/view_content/sidebar.tmpl:341]: Can we find another place for this button since to sidebar seems to overloaded to me. Maybe at the comment form?

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: models/issue_lock.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	return lockOrUnlockIssue(opts, CommentTypeLock)
+}
+
+// UnlockIssue unlocks a previously locked issue.
+func UnlockIssue(opts *IssueLockOptions) error {
+	opts.Issue.IsLocked = false
+	return lockOrUnlockIssue(opts, CommentTypeUnlock)
+}
+
+func lockOrUnlockIssue(opts *IssueLockOptions, commentType CommentType) error {
```
[the remark this answers is not in the document]
Comment (adelowo) [reply models/issue_lock.go]: Ha, makes sense.. I wanted to be as explicit as can be. Would fix

File: modules/auth/repo_form.go [hunk 6]
```diff
@@ -302,6 +303,32 @@ func (f *ReactionForm) Validate(ctx *macaron.Context, errs binding.Errors) bindi
...
+}
+
+// Validate validates the fields
+func (i *IssueLockForm) Validate(ctx *macaron.Context, errs binding.Errors) binding.Errors {
+	return validate(errs, ctx.Data, i, ctx.Locale)
+}
+
+// HasValidReason checks to make sure that the reason submitted in
+// the form matches any of the values in the config
+func (i IssueLockForm) HasValidReason() bool {
```
Comment (jonasfranz) [inline modules/auth/repo_form.go:324]: Maybe add this to `Validate`.

File: options/locale/locale_en-US.ini [hunk 7]
```diff
@@ -756,6 +756,24 @@ issues.attachment.open_tab = `Click to see "%s" in a new tab`
...
+issues.unlock_confirm = Unlock
+issues.lock.notice_1 = - Other users can’t add new comments to this issue.
+issues.lock.notice_2 = - You and other collaborators with access to this repository can still leave comments that others can see.
+issues.lock.notice_3 = - You can always unlock this issue again in the future.
+issues.unlock.notice_1 = - Everyone would be able to comment on this issue once more.
+issues.unlock.notice_2 = - You can always lock this issue again in the future.
+issues.lock.reason = Reason for locking
+issues.lock.title = Lock conversation on this issue.
+issues.unlock.title = Unlock conversation on this issue.
+issues.comment_on_locked = You cannnot comment on a locked issue.
```
Comment (jonasfranz) [inline options/locale/locale_en-US.ini]: typo (`cannnot`)

**Review by jonasfranz — DISMISSED** [review]
Please add a middleware to `router.go` in order to prevent creating comments via POST requests or API requests.

**Comment by adelowo** [comment]
> prevent creating comments 

You can still create comments but you must have write access, 
I attached the repoWriter middleware to the route already..
See https://github.com/go-gitea/gitea/pull/5073/files#diff-7aa71b75b5fddf1332a485052ffbca9aR543

and https://github.com/go-gitea/gitea/pull/5073/files#diff-ece22e0fe138da6e0fe0f25213406313R1033 

---- 

Would fix for api requests

File: modules/auth/repo_form.go (hunk 6 above)
Comment (adelowo) [reply modules/auth/repo_form.go:324]: I think it is better if explicit enough ( I have a couple merged PRs that follow this route though ) .. And that regular `Validate` is just meant for validating the struct bindings.. 

I wouldn't want to mess with `bindings.Error`

Comment (adelowo) [reply modules/auth/repo_form.go:324]: Thoughts @JonasFranzDEV ?

Comment (ghost) [reply modules/auth/repo_form.go:324]: Nudge

Comment (techknowlogick) [reply modules/auth/repo_form.go:324]: ping @jonasfranz

File: models/issue_lock.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (kolaente) [inline models/issue_lock.go]: Please update the year.

File: routers/repo/issue.go [hunk 9]
```diff
@@ -61,6 +61,23 @@ var (
 	}
 )
 
+// MustAllowUserComment checks to make sure if an issue is locked.
+// If locked and user has permissions to write to the repository,
```
Comment (kolaente) [inline routers/repo/issue.go:61]: What do you think about making it an option whether or not a repo writer can still comment on a locked issue?

Comment (adelowo) [reply routers/repo/issue.go:61]: Sounds nice.

Will this potentially be a global config or an option set at the point of ___locking the issue___ ?

Comment (kolaente) [reply routers/repo/issue.go:61]: I'd make a global option (in the config file) and then a setting in the repo to overwrite the global default.

File: custom/conf/app.ini.sample [hunk 10]
```diff
@@ -67,6 +67,10 @@ MAX_FILES = 5
 ; List of prefixes used in Pull Request title to mark them as Work In Progress
 WORK_IN_PROGRESS_PREFIXES=WIP:,[WIP]
 
+[repository.issue]
+; List of reasons why a Pull Request or Issue can be locked
```
Comment (kolaente) [inline custom/conf/app.ini.sample:73]: Please add this to the docs.

File: custom/conf/app.ini.sample [hunk 11]
```diff
@@ -336,8 +340,8 @@ ENABLE_CAPTCHA = false
 CAPTCHA_TYPE = image
 ; Enable recaptcha to use Google's recaptcha service
 ; Go to https://www.google.com/recaptcha/admin to sign up for a key
-RECAPTCHA_SECRET  =
-RECAPTCHA_SITEKEY =
+RECAPTCHA_SECRET  =
```
Comment (kolaente) [inline custom/conf/app.ini.sample]: Why did you change this?

**Review by kolaente — CHANGES_REQUESTED** [review]

File: custom/conf/app.ini.sample (hunk 11 above)
Comment (adelowo) [reply custom/conf/app.ini.sample]: Looks like a few spaces added there.. Will revert

**Comment by adelowo** [comment]
CI has been fixed @lunny

**Comment by adelowo** [comment]
Can anyone help me look at the icon thing ?  So it doesn't align because other icons have this `octicon-primitive-iconName` but what I am using doesn't ( `octicion-lock` and `octicon-key`) .. 

I have added an hack here though :) 
https://github.com/go-gitea/gitea/pull/5073/commits/428d8dfa12969aaeea5773f14cbdb5810fcd8e3 3

**Comment by lunny** [comment]
@adelowo conflicted again :(

**Comment by adelowo** [comment]
~~I’m having breakfast ... Will fix once I’m done :)~~

Fixed

File: models/repo_permission.go [hunk 12]
```diff
@@ -16,6 +16,12 @@ func (p *Permission) IsOwner() bool {
 	return p.AccessMode >= AccessModeOwner
 }
 
+// IsWriter checks if the current user has write or higher access to the
+// repository.
+func (p *Permission) IsWriter() bool {
+	return p.AccessMode >= AccessModeWrite
```
Comment (lunny) [inline models/repo_permission.go]: This method have been deleted since we should use `CanWrite(models.UnitTypeIssues)` instead here.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/repo_permission.go (hunk 12 above)
Comment (adelowo) [reply models/repo_permission.go]: @lunny Curiosity as I am not familiar with that part of the code. But do I need to use since both issues and PRs can be locked

```
CanWrite(models.UnitTypeIssues) && CanWrite(models.UnitTypePullRequests)
```

**Comment by adelowo** [comment]
This is ready for review now :)

File: modules/setting/setting.go [hunk 13]
```diff
@@ -290,6 +295,13 @@ var (
 		}{
 			WorkInProgressPrefixes: []string{"WIP:", "[WIP]"},
 		},
+
+		// Issue settings
+		Issue: struct {
+			LockReasons []string
+		}{
+			LockReasons: strings.Split("Too heated,Off-topic,Spam,Resolved", ","),
```
Comment (kolaente) [inline modules/setting/setting.go:292]: Where are these filled from `app.ini`?

File: routers/api/v1/repo/issue_comment.go [hunk 14]
```diff
@@ -169,6 +170,11 @@ func CreateIssueComment(ctx *context.APIContext, form api.CreateIssueCommentOpti
 		return
 	}
 
+	if issue.IsLocked && !ctx.Repo.CanWrite(models.UnitTypeIssues) && !ctx.User.IsAdmin {
+		ctx.Error(500, "CreateIssueComment", errors.New(ctx.Tr("repo.issues.comment_on_locked")))
```
Comment (kolaente) [inline routers/api/v1/repo/issue_comment.go]: This shouldn't be a 500 if the user does not have the right to do it. Forbidden would be better (403)

File: routers/repo/issue.go [hunk 15]
```diff
@@ -896,6 +913,11 @@ func ViewIssue(ctx *context.Context) {
 	ctx.Data["SignInLink"] = setting.AppSubURL + "/user/login?redirect_to=" + ctx.Data["Link"].(string)
 	ctx.Data["IsIssuePoster"] = ctx.IsSigned && issue.IsPoster(ctx.User.ID)
 	ctx.Data["IsIssueWriter"] = ctx.Repo.CanWriteIssuesOrPulls(issue.IsPull)
+	ctx.Data["IsRepoAdmin"] = ctx.IsSigned && ctx.Repo.IsAdmin() && ctx.User.IsAdmin
+	ctx.Data["IsWriter"] = ctx.IsSigned && ctx.Repo.CanWrite(models.UnitTypeIssues) && ctx.User.IsAdmin
+	ctx.Data["IsRepoAdmin"] = ctx.IsSigned && (ctx.Repo.IsAdmin() || ctx.User.IsAdmin)
```
Comment (kolaente) [inline routers/repo/issue.go:918]: Why two times `IsRepoAdmin`?

File: routers/repo/issue.go [hunk 16]
```diff
@@ -896,6 +913,11 @@ func ViewIssue(ctx *context.Context) {
...
+	ctx.Data["IsRepoWriter"] = ctx.IsSigned && (ctx.Repo.CanWrite(models.UnitTypeIssues) || ctx.User.IsAdmin)
```
Comment (kolaente) [inline routers/repo/issue.go]: What is the difference between `IsWriter` and `IsRepoWriter`? Why are both needed?

File: routers/repo/issue.go [hunk 17]
```diff
@@ -57,6 +57,23 @@ var (
...
+func MustAllowUserComment(ctx *context.Context) {
+
+	issue := GetActionIssue(ctx)
+	if ctx.Written() {
+		return
+	}
+
+	if issue.IsLocked && !ctx.Repo.CanWrite(models.UnitTypeIssues) && !ctx.User.IsAdmin {
+		ctx.Flash.Error(ctx.Tr("repo.issues.comment_on_locked"))
+		ctx.Redirect(fmt.Sprintf("%s/issues/%d", ctx.Repo.RepoLink, issue.Index))
```
Comment (kolaente) [inline routers/repo/issue.go]: Please use `ctx.Redirect(issue.HTMLURL())` as you did in other functions.

File: routers/routes/routes.go [hunk 18]
```diff
@@ -406,6 +406,13 @@ func RegisterRoutes(m *macaron.Macaron) {
 	reqRepoIssuesOrPullsWriter := context.RequireRepoWriterOr(models.UnitTypeIssues, models.UnitTypePullRequests)
 	reqRepoIssuesOrPullsReader := context.RequireRepoReaderOr(models.UnitTypeIssues, models.UnitTypePullRequests)
 
+	reqRepoWriter := func(ctx *context.Context) {
```
Comment (kolaente) [inline routers/routes/routes.go]: Could you change that to something like `RepoIssueWriter`? IMHO `RepoWriter` indicates the user can also push to the repo etc.
The name is somewhat misleading.

File: templates/repo/issue/view_content/comments.tmpl [hunk 19]
```diff
@@ -1,7 +1,11 @@
 {{range .Issue.Comments}}
 	{{ $createdStr:= TimeSinceUnix .CreatedUnix $.Lang }}
 
-	<!-- 0 = COMMENT, 1 = REOPEN, 2 = CLOSE, 3 = ISSUE_REF, 4 = COMMIT_REF, 5 = COMMENT_REF, 6 = PULL_REF, 7 = COMMENT_LABEL, 12 = START_TRACKING, 13 = STOP_TRACKING, 14 = ADD_TIME_MANUAL, 16 = ADDED_DEADLINE, 17 = MODIFIED_DEADLINE, 18 = REMOVED_DEADLINE, 19 = ADD_DEPENDENCY, 20 = REMOVE_DEPENDENCY, 21 = CODE, 22 = REVIEW -->
+	<!-- 0 = COMMENT, 1 = REOPEN, 2 = CLOSE, 3 = ISSUE_REF, 4 = COMMIT_REF,
+	 5 = COMMENT_REF, 6 = PULL_REF, 7 = COMMENT_LABEL, 12 = START_TRACKING,
+	 13 = STOP_TRACKING, 14 = ADD_TIME_MANUAL, 16 = ADDED_DEADLINE, 17 =
+	MODIFIED_DEADLINE, 18 = REMOVED_DEADLINE, 19 = ADD_DEPENDENCY, 20 =
```
Comment (kolaente) [inline templates/repo/issue/view_content/comments.tmpl]: Small formatting issue: could you not end new lines like this one with a `=`?

**Review by kolaente — CHANGES_REQUESTED** [review]

File: routers/repo/issue.go (hunk 16 above)
Comment (adelowo) [reply routers/repo/issue.go]: Turns out `IsWriter` was redundant and not used anywhere

**Review by lafriks — APPROVED** [review]

**Comment by adelowo** [comment]
> This will make a great addition to Gitea. I hope to have the ability to close _my own_ issues and delete _my own_ comments even if a conversation is locked – otherwise the need for a ToS arises fairly quickly.

Sure thing, you can still perform all other functions. The only thing you cannot do is to create a new comment.

File: templates/repo/issue/view_content/comments.tmpl [hunk 20]
```diff
@@ -354,5 +358,34 @@
...
+		<div class="event">
+			<span class="octicon octicon-lock"
+				style="font-size:20px;margin-left:-28.5px; margin-right: -1px"></span>
+			<a class="ui avatar image" href="{{.Poster.HomeLink}}">
+				<img src="{{.Poster.RelAvatarLink}}">
+			</a>
+
+			{{ $formatting := "" }}
+			{{ if .Content }}
+				{{ $formatting = "as " }}
```
Comment (lafriks) [inline templates/repo/issue/view_content/comments.tmpl]: This is untranslatable...

File: models/repo_permission.go (hunk 12 above)
Comment (lunny) [reply models/repo_permission.go]: @adelowo You should add a `if`:
```go
if issue.IsPull {
    return CanWrite(models. UnitTypePullRequests)
}
return CanWrite(models. UnitTypeIssues)
```

File: routers/repo/issue.go (hunk 15 above)
Comment (lunny) [reply routers/repo/issue.go:918]: `IsWriter` is duplicated with `IsIssueWriter`. And
```go
ctx.Data["IsRepoAdmin"] = ctx.IsSigned && (ctx.Repo.IsAdmin() || ctx.User.IsAdmin)
```

File: templates/repo/issue/view_content/comments.tmpl (hunk 20 above)
Comment (adelowo) [reply templates/repo/issue/view_content/comments.tmpl]: Done, you can take a look now

**Comment by lafriks** [comment]
@lunny @jonasfranz please review

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@jonasfranz Please review this PR.

**Comment by lafriks** [comment]
@kolaente please review

File: routers/repo/issue.go (hunk 9 above)
Comment (kolaente) [reply routers/repo/issue.go:61]: So, what about this? I think it'd be ok if we move the configuration to another pr.

**Review by kolaente — COMMENTED** [review]
Looks pretty good so far, just one tiny nit.

File: routers/repo/issue.go (hunk 9 above)
Comment (lafriks) [reply routers/repo/issue.go:61]: @kolaente this PR is already long history, let's move such option to other PR

Comment (kolaente) [reply routers/repo/issue.go:61]: @lafriks sounds good!

**Review by kolaente — APPROVED** [review]
